mm2src/mm2_main/src/lp_swap/maker_swap.rs

```diff
@@ -810,16 +818,24 @@ impl MakerSwap {
             })
             .compat();
 
-        let reward_amount = if self.r().watcher_reward {
-            let reward = match watcher_reward_amount(&self.maker_coin, &self.taker_coin).await {
-                Ok(reward) => reward,
+        let watcher_reward = if self.r().watcher_reward && self.maker_coin.is_eth() {
+            match get_watcher_reward(
+                &self.maker_coin,
+                &self.taker_coin,
+                Some(self.maker_amount.clone()),
+                Some(self.taker_amount.clone()),
+                false,
+                self.watcher_reward_amount(),
+            )
```

Comment:
I don't like having specific coin logic in swaps code, maybe we can use `validate_maker_payment_instructions`/ `validate_taker_payment_instructions` to save `WatcherReward` in `MakerSwapEvent::MakerPaymentInstructionsReceived` \ `TakerSwapEvent::TakerPaymentInstructionsReceived` https://github.com/KomodoPlatform/atomicDEX-API/blob/b47f95ecdd51f40692bbb6e0bed8142fa380b4d4/mm2src/mm2_main/src/lp_swap/maker_swap.rs#L723
We can change `validate_maker_payment_instructions` name to `validate_and_generate_maker_payment_instructions` or something more appropriate, same for `validate_taker_payment_instructions`.
we will add some more fields to `PaymentInstructionArgs`, it can be split into 2 structs `PaymentInstructionArgs`,`ValidatePaymentInstructionArgs` if needed.
What do you think about this approach? does it have any drawbacks that you can think of? Is there an alternative better way to avoid having specific coins logic here?

Reply:
We need the watcher reward in the `TakerFeeData` message, before the `validate_taker_payment_instructions` is called. Are you talking about the `self.maker_coin.is_eth()` part as coin specific? The reason for having that check here was to avoid calling an async function if the watcher reward is not going to be used after a discussion with @ozkanonur. I can move that part inside the `get_watcher_reward ` function to remove the coin specific checks from the `maker_swap` and `taker_swap`. Another thing I could do is to implement get_watcher_reward for each coin, and then use common functions again to avoid duplication.

Reply:
> Are you talking about the self.maker_coin.is_eth() part as coin specific?

Yes, that's what I meant, sorry if I was not clear. Also, there is a lot of coin specific code inside `get_watcher_reward` and the functions it uses in `lp_swaps.rs`.

> We need the watcher reward in the TakerFeeData message, before the validate_taker_payment_instructions is called.

Thanks for this info, I didn't notice it, now I see why validate instruction functions can't be used for this.

> Another thing I could do is to implement get_watcher_reward for each coin, and then use common functions again to avoid duplication.

This is the solution I like the most, there can be 2 functions `get_taker_watcher_reward`/`get_maker_watcher_reward`. The problem will be in the `fetch_swap_coins_price` since it can't be imported to the `coins` crate, we will need to move this to a common crate or to create a new `mm2_price` crate.

P.S.  Using `is_eth()` in a few places like now is not a big problem in itself, but having this function can encourage relying on it more in the future, making the swaps code and maybe other parts (e.g. ordermatching when implementing native swaps) rely on a lot of conditionals depending on coins instead of implementing a coin's specific logic where it belongs.

Reply:
There are still 2 instances of `is_eth` left in swaps related code
https://github.com/KomodoPlatform/atomicDEX-API/blob/52c95cd43359f5cc89a26404f5071d903e206842/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L944
https://github.com/KomodoPlatform/atomicDEX-API/blob/52c95cd43359f5cc89a26404f5071d903e206842/mm2src/mm2_main/src/lp_swap/swap_watcher.rs#L644
It's ok to leave them if there is no other way, just making sure you didn't miss those.

Reply:
> Or I can do it once in taker_swap swap_watcher and pass the resulting bool as it currently is.

Let's keep it as is since there is no way around it.